File: lib/add-issue.py
"""Record a filed GitHub issue in the FLOW state file.

Usage:
  bin/flow add-issue --label <label> --title <title> --url <url> --phase <phase>

Appends to the issues_filed array in the state file. Follows the same
pattern as append-note.py for state file discovery and mutation.

Output (JSON to stdout):
  Success:  {"status": "ok", "issue_count": N}
  No state: {"status": "no_state"}
  Error:    {"status": "error", "message": "..."}
"""

import argparse
import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

from flow_utils import now, project_root, resolve_branch, PHASE_NAMES


def add_issue(state_path, label, title, url, phase):
    """Append an issue to the state file. Returns the updated state dict."""
    state = json.loads(state_path.read_text())

    if "issues_filed" not in state:
        state["issues_filed"] = []

    state["issues_filed"].append({
        "label": label,
        "title": title,
        "url": url,
        "phase": phase,
        "phase_name": PHASE_NAMES.get(phase, phase),
        "timestamp": now(),
    })

    state_path.write_text(json.dumps(state, indent=2))
    return state


def main():
    parser = argparse.ArgumentParser(description="Record a filed issue in FLOW state")
    parser.add_argument("--label", required=True, help="Issue label (e.g. Rule, Flow, Flaky Test)")
    parser.add_argument("--title", required=True, help="Issue title")
    parser.add_argument("--url", required=True, help="Issue URL")
    parser.add_argument("--phase", required=True, help="Phase that filed the issue")
    parser.add_argument("--branch", type=str, default=None, help="Override branch for state file lookup")
    args = parser.parse_args()

    root = project_root()
    branch, candidates = resolve_branch(args.branch)

    if branch is None:
        if candidates:
            print(json.dumps({
                "status": "error",
                "message": "Multiple active features. Pass --branch.",
                "candidates": candidates,
            }))
        else:
            print(json.dumps({
                "status": "error",
                "message": "Could not determine current branch",
            }))
        sys.exit(1)

    state_path = root / ".flow-states" / f"{branch}.json"

    if not state_path.exists():
        print(json.dumps({"status": "no_state"}))
        sys.exit(0)

    try:
        json.loads(state_path.read_text())
    except Exception as e:
        print(json.dumps({
            "status": "error",
            "message": f"Could not read state file: {e}",
        }))
        sys.exit(1)

    try:
        state = add_issue(state_path, args.label, args.title, args.url, args.phase)
    except Exception as e:
        print(json.dumps({
            "status": "error",
            "message": f"Failed to add issue: {e}",
        }))
        sys.exit(1)

    print(json.dumps({
        "status": "ok",
        "issue_count": len(state["issues_filed"]),
    }))


if __name__ == "__main__":
    main()

File: src/add_issue.rs
//! `bin/flow add-issue` — record a filed issue in FLOW state.
//!
//! Tests live at `tests/add_issue.rs` per
//! `.claude/rules/test-placement.md` — no inline `#[cfg(test)]` in
//! this file.

use std::path::Path;

use clap::Parser;
use serde_json::{json, Value};

use crate::flow_paths::FlowPaths;
use crate::git::resolve_branch;
use crate::lock::mutate_state;
use crate::phase_config::phase_names;
use crate::utils::now;

#[derive(Parser, Debug)]
#[command(name = "add-issue", about = "Record a filed issue in FLOW state")]
pub struct Args {
    /// Issue label (e.g. Rule, Tech Debt, Documentation Drift)
    #[arg(long)]
    pub label: String,

    /// Issue title
    #[arg(long)]
    pub title: String,

    /// Issue URL
    #[arg(long)]
    pub url: String,

    /// Phase that filed the issue
    #[arg(long)]
    pub phase: String,

    /// Override branch for state file lookup
    #[arg(long)]
    pub branch: Option<String>,
}

/// Applies the issues_filed append transform to the in-memory state.
///
/// Extracted to a named function so cargo-llvm-cov measures a single
/// monomorphization of the mutation logic regardless of how many
/// tests or production paths call [`run_impl_main`]. The closure
/// passed to [`mutate_state`] becomes a thin delegator, while every
/// region inside `apply_issue_mutation` (object-guard early return,
/// missing-array auto-create, array push) reaches 100% merged
/// coverage under the test suite.
fn apply_issue_mutation(state: &mut Value, args: &Args, phase_name: &str, timestamp: &str) {
    // Corruption resilience: skip mutation when state root is wrong
    // type (e.g. array from interrupted write) to prevent IndexMut
    // panics. See .claude/rules/rust-patterns.md "State Mutation
    // Object Guards".
    if !(state.is_object() || state.is_null()) {
        return;
    }
    if state.get("issues_filed").is_none() || !state["issues_filed"].is_array() {
        state["issues_filed"] = json!([]);
    }
    // The block above guarantees state["issues_filed"] is an array,
    // so as_array_mut returns Some unconditionally.
    let arr = state["issues_filed"]
        .as_array_mut()
        .expect("issues_filed is always an array here");
    arr.push(json!({
        "label": args.label,
        "title": args.title,
        "url": args.url,
        "phase": args.phase,
        "phase_name": phase_name,
        "timestamp": timestamp,
    }));
}

/// Main-arm dispatcher with injected root. Returns `(value, exit_code)`:
/// `(ok+issue_count, 0)` on success, `(no_state, 0)` when the state file
/// is missing, `(error+message, 1)` on resolve-branch failure or
/// mutate_state failure. Tests pass tempdir paths and `--branch` args
/// to bypass git resolution.
pub fn run_impl_main(args: Args, root: &Path) -> (Value, i32) {
    let branch = match resolve_branch(args.branch.as_deref(), root) {
        Some(b) => b,
        None => {
            return (
                json!({"status": "error", "message": "Could not determine current branch"}),
                1,
            );
        }
    };
    // Branch reaches us either from `current_branch()` (raw git output)
    // or from `--branch` CLI override (raw user input). Both are
    // external inputs per `.claude/rules/external-input-validation.md`,
    // so use the fallible constructor to reject slash-containing or
    // empty branches as a structured error rather than a panic.
    let state_path = match FlowPaths::try_new(root, &branch) {
        Some(p) => p.state_file(),
        None => {
            return (
                json!({"status": "error", "message": format!("Invalid branch '{}'", branch)}),
                1,
            );
        }
    };

    if !state_path.exists() {
        return (json!({"status": "no_state"}), 0);
    }

    let names = phase_names();
    let phase_name = match names.get(&args.phase) {
        Some(n) => n.clone(),
        None => args.phase.clone(),
    };
    let timestamp = now();

    match mutate_state(&state_path, &mut |state| {
        apply_issue_mutation(state, &args, &phase_name, &timestamp);
    }) {
        Ok(state) => {
            let count = match state["issues_filed"].as_array() {
                Some(a) => a.len(),
                None => 0,
            };
            (json!({"status": "ok", "issue_count": count}), 0)
        }
        Err(e) => (
            json!({"status": "error", "message": format!("Failed to add issue: {}", e)}),
            1,
        ),
    }
}
